### ISSUE TITLE
SRU

### ISSUE THREAD

**Issue by edmondja** [keras-team/keras#7870]
Does somebody has projected to implement https://github.com/taolei87/sru SRU recurrent neural networks in Keras ?

**Comment by titu1994** [comment]
@edmondja Doesn't this require custom CUDA operations that are provided in the pytorch repo? 

If that is the case, while the equations can be written in Keras format, the performance of the implementation will severely lag behind the CUDA optimized version.

**Comment by ushakov** [comment]
The paper reports a 5x speedup over the normal LSTM even without the specialized CUDA code, so this may be worth adding anyway.

**Comment by edmondja** [comment]
Yes, cuda isnt needed. Plus you can mimic the cuda implementation with keras if im not wrong.

**Comment by titu1994** [comment]
Hmm interesting. 

It has a wierd Length first representation in contrast to Keras Batch First data shape, most probably for the optimized cuda algorithm to have efficient parallelization. 

Let's see if I can get it to work. No promises though.

**Comment by titu1994** [comment]
Seems it is not going to be straightforward to implement. The main issue is the fact that the final update for `h` requires the input `x_t`. However, even if I pass the input `x` itself as a constant to the step function, I cannot access just 1 sample from the batch.

Since the entire batch is passed, and there is no way to obtain the index of iteration inside the Tensorflow while loop, I can't access the ith sample of input. 

@fchollet Since this time step is available inside K.rnn(...), would there be any possible way to obtain and use that constant at each timestep directly to index the ith image in the loop over timesteps ?

**Comment by titu1994** [comment]
On second thought, I don't think I'll be able to implement this. 

There are other severe problems, such as the input shape mismatch (when input dim != output dim) and performance being only slightly better (17 seconds per epoch instead of 25 seconds using implementation=2, 128 batchsize and setting the index for input to 0 (its wrong but I just wanted to try out the speed)) without the cuda optimizations. In contrast, CNNs run through IMDB in around 7-8 seconds per epoch on my laptop.

I will post a gist of the incorrect partial version, hopefully someone can pick that up and get it to work eventually. https://gist.github.com/titu1994/638d0a571d368f4fdc8f8ca5eddd25cd

**Comment by titu1994** [comment]
Was able to implement the paper finally. https://github.com/titu1994/keras-SRU

While I fixed most of the issues I had before, the problem of speed remains. No more need to unroll the RNN (though that obviously helps with speed), and no problem accepting different input dimension than the output dimension. 

The only thing left is the speed problem. Sadly, I cant think of any way to make this faster without the custom cuda kernel being implemented in Tensorflow / Theano / CNTK. As can be seen, they also use the naive implementation for CNTK, since there seems to be some issue in implementing the cuda kernel there..

**Comment by edmondja** [comment]
Sorry for the late answer. 
Wow, very good job thanks !
Even if Keras and the backends have some limitations, I think its already very interesting.

**Comment by longzaitianguo** [comment]
@titu1994 how can I implement bi-sru using your sru cell?

**Comment by titu1994** [comment]
It shouldn't be any different than normal. Treat the sru layey same as an LSTM.
